### PR TITLE
feat(ui): Moaaar filters: add `all`, `any`, `not`, `unread` and `category` filters

### LINKED ISSUE #3020: [RoomList Filters] Create the "Rooms" and the "People" filters
We should create a filter for these two filters which should be mutually exclusive, since one is for group chats and the other for DMs.
These two filters are mutually exclusive and shouldn't be chained together, but can be chained with other filters

### LINKED ISSUE #3022: [RoomList Filters] Create the "Unreads" Filter
Create the filter for "Unreads" which should filter all the rooms with unread notifications (if unread mentions should be included is TBD).
This filter can be chained with other filters.

### PR DESCRIPTION
**Pull request by Hywan**
This PR adds more filters for the `matrix_sdk_ui::room_list_service`:

* [x] `all`
* [x] `any`
* [x] `not`
* [x] `unread`
* [x] `category`.

This patch removes the previous `all` filter. The `non_left` filter should be used instead.
This patch adds a **new** `all` filter, which is a “logical” all, just like `any` or `not`. All the 3 new filters must be used to _combine_ filters together, e.g (in pseudo-code):

* `all(non_left, category(people))`: all `non_left` rooms where the category of the room is `people`.
* `all(non_left, category(group), normalized_match_room_name("foo"))`: all `non_left` rooms where the category of the room is `group` and where the name matches `"foo"`.
* `all(non_left, unread, any(normalized_match_room_name("foo"), not(normalized_match_room_name("bar"))))`: all `non_left` rooms that contain unread messages and where the room name matches `"foo"` and does **not** match `"bar"`.

---

* Closes #3020
* Closes #3022

### CODE REVIEW

File: crates/matrix-sdk-ui/src/room_list_service/filters/unread.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    fn matches(&self, room_list_entry: &RoomListEntry) -> bool {
+        if !matches!(room_list_entry, RoomListEntry::Filled(_) | RoomListEntry::Invalidated(_)) {
+            return false;
+        }
+
+        let Some(read_receipts) = (self.room_read_receipts)(room_list_entry) else {
+            return false;
+        };
+
+        read_receipts.num_unread > 0
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk-ui/src/room_list_service/filters/unread.rs:26]: Answer is here https://github.com/element-hq/element-meta/issues/2270#issuecomment-1927252077

File: bindings/matrix-sdk-ffi/src/room_list.rs [hunk 2]
```diff
@@ -417,13 +410,60 @@ impl RoomListDynamicEntriesController {
...
+    Unread,
+    Category { expect: RoomListFilterCategory },
     None,
     NormalizedMatchRoomName { pattern: String },
     FuzzyMatchRoomName { pattern: String },
 }
 
+#[derive(uniffi::Enum)]
+pub enum RoomListFilterCategory {
+    Group,
```
Comment (andybalaam) [inline bindings/matrix-sdk-ffi/src/room_list.rs:425]: People is plural, so maybe Group should be Groups?

File: crates/matrix-sdk-ui/src/room_list_service/filters/any.rs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        let room_list_entry = RoomListEntry::Filled(room_id!("!r0:bar.org").to_owned());
+
+        {
+            let filter = |_: &_| true;
+            let any = new_filter(vec![Box::new(filter)]);
+
+            assert!(any(&room_list_entry));
+        }
+
+        {
```
Comment (andybalaam) [inline crates/matrix-sdk-ui/src/room_list_service/filters/any.rs:29]: I would prefer separate tests with a different name for each one, but this is just a preference.

File: crates/matrix-sdk-ui/src/room_list_service/filters/category.rs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        }
+
+        let kind = match (self.number_of_direct_targets)(room_list_entry) {
+            // If 1, we are sure it's a direct room between two users. It's the strict
+            // definition of the `People` category, all good.
+            Some(1) => RoomCategory::People,
+
+            // If smaller than 1, we are not sure it's a direct room, it's then a `Group`.
+            // If greater than 1, we are sure it's a direct room but not between
+            // two users, so it's a `Group` based on our expectation.
```
Comment (andybalaam) [inline crates/matrix-sdk-ui/src/room_list_service/filters/category.rs:44]: I assume we know that this is totally different from Element Web's idea of whether a room is a DM?

**Review by andybalaam — APPROVED** [review]
Looks great, sorry for the slowness on reviewing!

A couple of comments and questions, all optional.

File: bindings/matrix-sdk-ffi/src/room_list.rs [hunk 5]
```diff
@@ -417,13 +410,60 @@ impl RoomListDynamicEntriesController {
 
 #[derive(uniffi::Enum)]
 pub enum RoomListEntriesDynamicFilterKind {
-    All,
-    AllNonLeft,
+    All { filters: Vec<RoomListEntriesDynamicFilterKind> },
+    Any { filters: Vec<RoomListEntriesDynamicFilterKind> },
+    NonLeft,
+    Unread,
+    Category { expect: RoomListFilterCategory },
```
[the remark this answers is not in the document]
Comment (Hywan) [reply bindings/matrix-sdk-ffi/src/room_list.rs:417]: I thought about “value” initially but I did not convince me that it was the thing the room will be compared/tested against. I believe “expect” conveys this idea better. Thoughts?

File: bindings/matrix-sdk-ffi/src/room_list.rs (hunk 2 above)
Comment (Hywan) [reply bindings/matrix-sdk-ffi/src/room_list.rs:425]: Well yeah, I had the same thought :-D! A room contains a group, or multiple people, that's my reasoning.

File: crates/matrix-sdk-ui/src/room_list_service/filters/category.rs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+            // Don't know.
+            None => return false,
+        };
+
+        kind == expected_kind
+    }
+}
+
+/// Create a new filter that will accept all filled or invalidated entries, but
+/// filters out rooms that have no unread messages.
```
[the remark this answers is not in the document]
Comment (Hywan) [reply crates/matrix-sdk-ui/src/room_list_service/filters/category.rs:56]: Indeed, good eye!

File: crates/matrix-sdk-ui/src/room_list_service/filters/any.rs (hunk 3 above)
Comment (Hywan) [reply crates/matrix-sdk-ui/src/room_list_service/filters/any.rs:29]: :+1:

File: crates/matrix-sdk-ui/src/room_list_service/filters/category.rs (hunk 4 above)
Comment (Hywan) [reply crates/matrix-sdk-ui/src/room_list_service/filters/category.rs:44]: I totally disagree with the fact we need to count the number of participants in a room to decide whether it should be part of the People Category. I'm not doing the rule here :-(.

File: bindings/matrix-sdk-ffi/src/room_list.rs (hunk 2 above)
Comment (andybalaam) [reply bindings/matrix-sdk-ffi/src/room_list.rs:425]: But rooms inside People generally contain a person, right? (If you exclude yourself...)

Element Web has "People" and "Rooms". But I am happy with whichever you prefer.

File: bindings/matrix-sdk-ffi/src/room_list.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (andybalaam) [reply bindings/matrix-sdk-ffi/src/room_list.rs:417]: Yeah, I see the argument for expect. Let's keep it.

**Pull request merged**